yt/config.py

```diff
@@ -66,68 +73,164 @@
+            raise KeyError(f"Could not find {section}, {path} in configuration.")
+
+        return node
+
+    def update(self, new_values):
+        def copy_helper(dict_a, dict_b):
+            # Copies entries form dict_a in dict_b, inplace
+            for k in dict_a.keys():
+                if k not in dict_b:
+                    dict_b[k] = dict_a[k]
+                    continue
+                if isinstance(dict_a[k], dict):
+                    copy_helper(dict_a[k], dict_b[k])
+                else:
+                    dict_b[k] = dict_a[k]
+
+        copy_helper(new_values, self.values)
+
+    def has_section(self, section):
+        return section in self.values
+
+    def add_section(self, section):
+        if not self.has_section(section):
+            self.values[section] = {}
+
+    def __setitem__(self, key, value):
+        section, *option_path, option_name = key
+        if not self.has_section(section):
+            raise KeyError
+
+        node = self.values[section]
+        for p in option_path:
+            if p not in node:
+                node[p] = {}
+            node = node[p]
+
+        node[option_name] = value
+
+    def set(self, section, option, value):
+        if not isinstance(option, (tuple, list)):
+            option = (option,)
+        self[(section, *option)] = value
+
+    def read(self, file_names):
+        if not isinstance(file_names, (tuple, list)):
+            file_names = (file_names,)
+
+        file_names_read = []
+        for fname in file_names:
+            if not os.path.exists(fname):
+                continue
+            self.update(toml.load(fname))
+            file_names_read.append(fname)
+
+        return file_names_read
+
+    def write(self, fd):
+        def cleaner(d):
+            new_d = {}
+            for k, v in d.items():
+                if isinstance(v, dict):
+                    if len(v) > 0:
+                        new_d[k] = cleaner(v)
+                else:
+                    new_d[k] = v
+            return new_d
 
-    def __getitem__(self, key):
-        self.get(key[0], key[1])
+        # Clean up the sections
+        cleaned_values = cleaner(self.values)
+        config_as_str = toml.dumps(cleaned_values)
 
-    def get(self, section, option, *args, **kwargs):
-        val = super(YTConfigParser, self).get(section, option, *args, **kwargs)
-        return os.path.expanduser(os.path.expandvars(val))
+        try:
+            fd.write(config_as_str)
+        except AttributeError:
+            with open(fd, mode="w") as fdd:
+                fdd.write(config_as_str)
 
 
-ytcfg = YTConfigParser(ytcfg_defaults)
-ytcfg.read([_OLD_CONFIG_FILE, CURRENT_CONFIG_FILE, "yt.cfg"])
-if not ytcfg.has_section("yt"):
-    ytcfg.add_section("yt")
+# Walk the tree up until we find a config file
+ytcfg = YTConfig(ytcfg_defaults)
+if os.path.exists(CURRENT_CONFIG_FILE):
+    ytcfg.read(CURRENT_CONFIG_FILE)
+cwd = Path(".").absolute()
+while cwd.parent != cwd:
```

Comment:
I don't get how this loop is supposed to stop.

Reply:
It stops when you reach the root, since the parent of the root is the root itself (i.e. `Path("/") == Path("/").parent`). AFAIK this is only true when you reach the root. I'm curious how this is handled on WIndows though.

Reply:
Cool, I didn't know about this !